Unify docs about .mjs, .cjs, .js and .json config files (#2150)
babel/website: docs/config-files.md

Babel has two parallel config file formats, which can be used together, or independently.

* Project-wide configuration
  * `babel.config.json` files, with the different extensions
* File-relative configuration
  * `.babelrc.json` files, with the different extensions
  * `package.json` files with a `"babel"` key

## Project-wide configuration

New in Babel 7.x, Babel has a concept of a ["root"](options.md#root) directory, which defaults
to the current working directory. For project-wide configuration, Babel will automatically search

> for a `babel.config.json` file, or an equivalend one using the [supported extensions](#supported-file-extensions),

Reviewer: ```suggestion
for a `babel.config.json` file, or an equivalent one using the [supported extensions](#supported-file-extensions),
```

Project-wide configs can also be disabled by setting ["configFile"](options.md#configfile) to `false`.

## File-relative configuration

> Babel loads `.babelrc.json` files, or an equivalend one using the [supported extensions](#supported-file-extensions), by searching up the

Reviewer: ```suggestion
Babel loads `.babelrc.json` files, or an equivalent one using the [supported extensions](#supported-file-extensions), by searching up the
```

could require setting ["babelrcRoots"](options.md#babelrcroots).

Given that, it may be more desirable to rename the `.babelrc` to be a
[project-wide "babel.config.json"](#project-wide-configuration). As mentioned in the project-wide
section above, this may then require explicitly setting ["configFile"](options.md#configfile)
since Babel will not find the config file if the working directory isn't correct.

## Supported file extensions

Babel can be configured using any file extension natively supported by Node.js: you can use `.json`,
`.js`, `.cjs` and `.mjs`, both for `babel.config.json` and `.babelrc.json` files.

- `babel.config.json` and `.babelrc.json` are parsed as JSON5 and should contain an object matching
  the [options](options.md) format that Babel accepts.
  

>   It is recommended to choose this file type wherever possible: JS config files are

Reviewer: ```suggestion
  We recommend using this file type wherever possible: JS config files are
```

  handy if you have complex configuration that is conditional or otherwise computed at build time.
  However, the downside is that JS configs are less statically analyzable, and therefore have

>   negative effects on cacheability, linting, IDE autocomplete, etc.

Reviewer: Sidenote, we might later want to illustrate more concrete examples of the effects (i.e. if you use .json with ____ then performance is much slower, etc).

  Since `babel.config.json` and `.babelrc.json` are static JSON files, it allows other tools that
  use Babel such as bundlers to cache the results of Babel safely, which can be a huge build
  performance win.

- `babel.config.cjs` and `.babelrc.cjs` allow you to define your configuration as CommonJS,
  using `module.exports`.

> - `babel.config.mjs` and `.babelrc.mjs` use native ECMAScript modules. They are only supported by

Reviewer: They are supported by Node.js 13.2+ (or older versions via the `--experimental-modules` flag)

  Node.js 13.2 or newer, and might work in older versions when passing `--experimental-modules` flag
  to Node.js.
  Please remember that native ECMAScript modules are asynchronous (that's why `import()` always
  returns a promise!): for this reason, `.mjs` config files will throw when calling Babel
  synchronously.

- `babel.config.js` and `.babelrc.js` behave like the `.mjs` equivalents when your `package.json`
  file contains the [`"type": "module"`](https://nodejs.org/api/esm.html#esm_code_package_json_code_code_type_code_field)
  option, otherwise they are exactly the same as the `.cjs` files.

JavaScript configuration files can either export an object, or a function that when called will
return the generated configuration.
Function-returning configs are given a few special powers because they can access an API exposed
by Babel itself. See [Config Function API](#config-function-api) for more information.

> > For compatibility reasons, you could find `.babelrc` used as an alias for `.babelrc.json`.

Reviewer: ```suggestion
> For compatibility reasons, `.babelrc` is an alias for `.babelrc.json`.
```
